cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp

```diff
@@ -0,0 +1,409 @@
+  // already allowed or has higher priority that the currently allowed
+  if ((cmd_actuator_subs.allowed == VACANT) || (cmd_actuator_subs.allowed == idx) ||
+      (cmd_actuator_subs[idx]->priority > cmd_actuator_subs[cmd_actuator_subs.allowed]->priority))
+  {
+    if (cmd_actuator_subs.allowed != idx)
+    {
+      cmd_actuator_subs.allowed = idx;
+    }
+    return true;
+  }
+  return false;
+}
+
+void CmdActuatorMuxNodelet::finAnglesTimerCallback(const ros::TimerEvent& event, unsigned int idx)
+{
+  timerCallbackProcess(cmd_fin_angles_subs, idx, fin_angles_active_subscriber);
+}
+
+void CmdActuatorMuxNodelet::setRPMTimerCallback(const ros::TimerEvent& event, unsigned int idx)
+{
+  timerCallbackProcess(cmd_set_rpm_subs, idx, set_rpm_active_subscriber);
+}
+
+void CmdActuatorMuxNodelet::timerCallbackProcess(CmdActuatorSubscribers& cmd_actuator_subs,
+                                                 const unsigned int& idx,
+                                                 const ros::Publisher& active_subscriber)
+{
+  if (cmd_actuator_subs.allowed == idx ||
+      (idx == GLOBAL_TIMER && cmd_actuator_subs.allowed != VACANT))
+  {
+    if (idx == GLOBAL_TIMER)
+    {
+      // No messages timeout happened for ANYONE, so last active source got stuck without further
+      // messages; not a big problem, just dislodge it; but possibly reflect a problem in the
+      // controller
+      NODELET_WARN("Actuator : No messages from ANY input received");
+      NODELET_WARN("Actuator: %s dislodged due to general timeout",
+                   cmd_actuator_subs[cmd_actuator_subs.allowed]->name.c_str());
+    }
+
+    // No messages timeout happened to currently active source, so...
+    cmd_actuator_subs.allowed = VACANT;
+
+    // ...notify the world that nobody is publishing; its vacant
+    std_msgs::StringPtr acv_msg(new std_msgs::String);
+    acv_msg->data = "idle";
+    active_subscriber.publish(acv_msg);
+  }
+
+  if (idx != GLOBAL_TIMER) cmd_actuator_subs[idx]->active = false;
+}
+
+void CmdActuatorMuxNodelet::onInit()
+{
+  nh = this->getNodeHandle();
+  pnh = this->getPrivateNodeHandle();
+
+  /*********************
+  ** Dynamic Reconfigure
+  **********************/
+  dynamic_reconfigure_cb = boost::bind(&CmdActuatorMuxNodelet::reloadConfiguration, this, _1, _2);
+  dynamic_reconfigure_server =
+      new dynamic_reconfigure::Server<cmd_actuators_mux::reloadConfig>(pnh);
+  dynamic_reconfigure_server->setCallback(dynamic_reconfigure_cb);
+
+  fin_angles_active_subscriber = pnh.advertise<std_msgs::String>("fin_angle_active", 1, true);
+  set_rpm_active_subscriber = pnh.advertise<std_msgs::String>("set_rpm_active", 1, true);
+
+  // Notify the world that by now nobody is publishing on yet
+  std_msgs::StringPtr active_msg(new std_msgs::String);
+  active_msg->data = "idle";
+  fin_angles_active_subscriber.publish(active_msg);
+  set_rpm_active_subscriber.publish(active_msg);
```

Comment:
@patapufate meta: I don't think these two publisher will ever send mismatching values (i.e. we made sure both fins and thruster pay attention to the same source). In which case, why have two?

Reply:
Updated in 971573c